### DISCUSSION TITLE
cpp/basics/define_and_inline

### DISCUSSION THREAD

**Discussion by giscus** [imarvinle/CSGuide#50] [Announcements]
# cpp/basics/define_and_inline

编程指北: 计算机系统学习指南, 操作系统, 计算机网络, C++, Java, 算法, 数据结构

https://csguide.cn/cpp/basics/define_and_inline.html

**Comment by dadaguai666** [comment]
最后那里x++不会运行两次。对于(x++) > 10  ? (x++) : 10，x = 1，因为1 < 10，所以结果会返回10，后面的x++不会运行。

**Comment by RosaryXiang** [comment]
#define MAX(a, b) ((a) > (b) ? (a) : (c))

int x = 1;
int y = MAX(x++, 10);
// 宏定义 x++ 会被执行两次, 因为 x++ 会被替换到两个地方，使用内联函数则不会出现这个问题
这里不太对，x++只会被执行一次，因为x自增以前为1<10，（a）不会被执行

**Comment by Milchstrape** [comment]
#include <iostream>
using namespace std;
#define MAX(a, b) ((a) > (b) ? (a) : (b))


// 宏定义 x++ 会被执行两次, 因为 x++ 会被替换到两个地方，使用内联函数则不会出现这个问题
int main() {
    int x = 3;
    int y = MAX(x++, 1);
    cout<<x<<endl<<y;
}
这样才是被执行了两次，执行第一次后x=4>1所以y=4,执行第二次后x=5

**Comment by yumimanji** [comment]
宏是预处理阶段替换的呀, 怎么能是编译阶段呢
